### DISCUSSION TITLE
課題やってみたスレ

### DISCUSSION THREAD

**Discussion by ymmt2005** [ymmt2005/grpc-tutorial#4] [General]
どの言語でもいいので、課題をやってみたら成果をこちらで共有してください 😄

**Comment by satoru-takeuchi** [comment]
deadlineつけてのinfer呼び出しまで行きました。

https://github.com/satoru-takeuchi/deepthought/tree/infer2

queryの種類とdeadlineをランダムにして500ms間隔で10秒間queryを発行しました。以下出力例

```
./client 127.0.0.1:13333
query "Universe" with deadline(now + 700 ms)
couldn't get answer: "It would take longer"
query "Life" with deadline(now + 800 ms)
answer to "Life" is 42
Boot: I THINK THEREFORE I AM.
query "Universe" with deadline(now + 700 ms)
couldn't get answer: "It would take longer"
query "Everyting" with deadline(now + 500 ms)
couldn't get answer: "Contemplate your query"
query "Universe" with deadline(now + 500 ms)
couldn't get answer: "It would take longer"
query "Everyting" with deadline(now + 700 ms)
couldn't get answer: "Contemplate your query"
query "Universe" with deadline(now + 800 ms)
Boot: I THINK THEREFORE I AM.
answer to "Universe" is 42
query "Life" with deadline(now + 800 ms)
Boot: I THINK THEREFORE I AM.
answer to "Life" is 42
query "Everyting" with deadline(now + 700 ms)
couldn't get answer: "Contemplate your query"
query "Universe" with deadline(now + 900 ms)
Boot: I THINK THEREFORE I AM.
answer to "Universe" is 42
boot canceled
query "Life" with deadline(now + 700 ms)
finished quer
```

**Reply by ymmt2005** [reply]
Congratulations!

**Reply by satoru-takeuchi** [reply]
プロトコルを拡張して前方/後方互換性があることを確認しました。

https://github.com/satoru-takeuchi/deepthought/releases/tag/compat

- 前方互換性の確認: 古いserverへ新しいclientからアクセス
  - Boot requestにおいてSilentに1を与えても単に無視されたので影響が無い

- 後方互換性の確認: 新しいserverへ古いclientからアクセス
  - Boot requestにおいてSilentパラメタはなくてもいいので問題ない
  - Infer requestへのresponseにおいてdescriptionが""で返ってきた。とくに内容に意味を持たせていないので問題ない

**Reply by satoru-takeuchi** [reply]
keepaliveの設定をしました。

https://github.com/satoru-takeuchi/deepthought/releases/tag/keepalive

以下その過程でわかったことを共有します。

```
> failed to clone data: failed to clone data from 10.x.y.z: rpc error: code = Unknown desc = closing transport due to: connection error: desc = "error reading from server: EOF", received prior goaway: code: ENHANCE_YOUR_CALM, debug data: too_many_pings
```

上記メッセージを確認するために`keepalive.ClientParameters.Time`を1msにしてから`Infer()`を実行しましたが、エラーは発生しませんでした。パラメタのコメントを見た結果、この値は10秒以未満にはできないことを知りました。

その後、ping間隔を10秒にした後に`Infer()`の所要時間を21秒にしてもエラーが起きませんでした。gRPCのソースを確認したところ、デフォルトでは初回のpingフレームを送った後にpingフレームを閾値である5分を待たずに送信、ということを3回繰り返すと上記のエラーに至ることがわかりました。

ref.
https://github.com/grpc/grpc-go/blob/c513103bee39e1ebc3793e7128941794667779de/internal/transport/http2_server.go#L818-L824

この情報をもとに以下PRを作成しました(マージ済)。
https://github.com/ymmt2005/grpc-tutorial/pull/6

**Comment by satoru-takeuchi** [comment]
全部終えました。コンテンツの提供ありがとうございました。
